plugins/sinks/shield/sink.go

```diff
@@ -0,0 +1,207 @@
+package shield
+
+import (
+	"context"
+	_ "embed"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/MakeNowJust/heredoc"
+	"github.com/odpf/meteor/models"
+	assetsv1beta2 "github.com/odpf/meteor/models/odpf/assets/v1beta2"
+	"github.com/odpf/meteor/plugins"
+	"github.com/odpf/meteor/registry"
+	"github.com/odpf/salt/log"
+	sh "github.com/odpf/shield/proto/v1beta1"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/types/known/anypb"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+//go:embed README.md
+var summary string
+
+type Config struct {
+	Host    string            `mapstructure:"host" validate:"required"`
+	Headers map[string]string `mapstructure:"headers"`
+}
+
+var info = plugins.Info{
+	Description: "Send user information to shield grpc service",
+	Summary:     summary,
+	Tags:        []string{"grpc", "sink"},
+	SampleConfig: heredoc.Doc(`
+	# The hostname of the shield service
+	host: https://shield.com
+	# Additional headers send to shield, multiple headers value are separated by a comma
+	headers:
+	  X-Shield-Email: meteor@odpf.io
+      X-Other-Header: value1, value2
+	`),
+}
+
+type Sink struct {
+	plugins.BasePlugin
+	client Client
+	config Config
+	logger log.Logger
+}
+
+func New(c Client, logger log.Logger) plugins.Syncer {
+	s := &Sink{
+		logger: logger,
+		client: c,
+	}
+	s.BasePlugin = plugins.NewBasePlugin(info, &s.config)
+
+	return s
+}
+
+func (s *Sink) Init(ctx context.Context, config plugins.Config) error {
+	if err := s.BasePlugin.Init(ctx, config); err != nil {
+		return err
+	}
+
+	if err := s.client.Connect(ctx, s.config.Host); err != nil {
+		return fmt.Errorf("error connecting to host: %w", err)
+	}
+
+	return nil
+}
+
+func (s *Sink) Sink(ctx context.Context, batch []models.Record) error {
+	for _, record := range batch {
+		metadata := record.Data()
+		s.logger.Info("sinking record to shield", "record", metadata.GetUrn())
+
+		shieldPayload, err := s.buildShieldPayload(metadata)
+		if err != nil {
+			return errors.Wrap(err, "failed to build shield payload")
+		}
+		if err = s.send(ctx, shieldPayload); err != nil {
+			return errors.Wrap(err, "error sending data")
+		}
+
+		s.logger.Info("successfully sinked record to shield", "record", metadata.Name)
+	}
+
+	return nil
+}
+
+func (s *Sink) Close() (err error) {
+	return
+	//TODO: Connection closes even when some records are unpiblished
+	//TODO: return s.client.Close()
```

Comment:
you can call `s.client.Close()` in this hook

Reply:
Can't do this, as this leads to the early closing of the connection.
This is also mentioned [here](https://github.com/odpf/meteor/blob/9de6002d1abd7c8ca2ee0f631a4d6ab8c46014a6/agent/agent.go#L283)